### PR TITLE
Fixed multiple editor resource renders interfering with each other 

### PR DESCRIPTION
**Pull request by Mailaender**
Followup of https://github.com/OpenRA/OpenRA/pull/16710.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Needs a rebase already. :(

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs [hunk 1]
```diff
@@ -219,8 +164,5 @@ public struct EditorCellContents
 		public static readonly EditorCellContents Empty = default(EditorCellContents);
 		public ResourceType Type;
 		public int Density;
-		public string Variant;
-		public ISpriteSequence Sequence;
-		public int Frame;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:224]: With the removal of these properties we can use the `CellContents` struct and remove the `EditorCellContents` struct.

**Review by teinarss — CHANGES_REQUESTED** [review]

File: OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs (hunk 1 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:224]: Done.

File: OpenRA.Mods.D2k/Traits/World/D2kEditorResourceRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+	public class D2kEditorResourceRenderer : EditorResourceRenderer
+	{
+		readonly Actor self;
+
+		public D2kEditorResourceRenderer(Actor self, D2kEditorResourceRendererInfo info)
+			: base(self, info)
+			{
+				this.self = self;
+			}
```
Comment (teinarss) [inline OpenRA.Mods.D2k/Traits/World/D2kEditorResourceRenderer.cs]: Wrong indentation?

File: OpenRA.Mods.D2k/Traits/World/D2kEditorResourceRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		{
+			return D2kResourceRenderer.Variants.Keys.Random(Game.CosmeticRandom);
+		}
+
+		bool CellContains(CPos c, ResourceType t)
+		{
+			return self.World.Map.Contains(c) && ResourceLayer.GetResourceType(c) == t;
+		}
+
+		ClearSides FindClearSides(ResourceType t, CPos p)
```
Comment (teinarss) [inline OpenRA.Mods.D2k/Traits/World/D2kEditorResourceRenderer.cs:75]: Now we have the same logic in 3 places, could be worth to extract this?

Comment (Mailaender) [reply OpenRA.Mods.D2k/Traits/World/D2kEditorResourceRenderer.cs:75]: I can't come up with something smart. The `CellContains` function is different.

**Comment by teinarss** [comment]
Regression: When painting overlays in map editor it produce less dense fields. Red arrows marks fields made in this PR.
image

**Comment by Mailaender** [comment]
Fixed it.

**Comment by pchote** [comment]
Can you please pull the changes from 38a5825ec03f2e806f631f32257a64968c5b71de into this PR?

This addresses several review comments:
* Forgot to remove `D2kEditorResourceLayer` and add the update rule to the update path.
* Removes redundant checks and CellChanged notifications.
* Simplifies and restructures code to prepare for the density-from-neighbours logic becoming optional in my vein PR.

**Comment by pchote** [comment]
The `*EditorResourceRenderer`s looks to be an almost identical copy of `*ResourceRenderer`s, changing only the trait references and visibility checks. We should be able to avoid this massive duplication by adding a new interface that abstracts the differences between those two traits and use the existing `*ResourceRenderer`s in the editor too!

**Comment by pchote** [comment]
I have finally filed #18515 that implements my request above, so lets continue this there. This PR was extremely helpful in first highlighting the problem and then prototyping ways to solve it. Its a shame that we didn't think to go with the newer solution right from the start!

**Pull request closed without merging**